### PR TITLE
feat: add health check to vault

### PR DESCRIPTION
**Pull request by steffenix**
Adds the health checks to the vaults.

### CODE REVIEW

File: contracts/BaseStrategy.sol [hunk 1]
```diff
@@ -16,6 +16,9 @@ struct StrategyParams {
     uint256 totalDebt;
     uint256 totalGain;
     uint256 totalLoss;
+    bool enforeChangeLimit;
```
Comment (enie123) [inline contracts/BaseStrategy.sol:19]:
```suggestion
    bool enforceChangeLimit;
```
nit: typo

File: contracts/Vault.vy [hunk 2]
```diff
@@ -1297,6 +1310,32 @@ def updateStrategyPerformanceFee(
     log StrategyUpdatePerformanceFee(strategy, performanceFee)
 
 
+@external
+def setStrategyEnforeChangeLimit(strategy: address, enabled: bool):
```
Comment (enie123) [inline contracts/Vault.vy:1314]: nit: typo `Enfore` -> `Enforce` in a few places

File: contracts/Vault.vy [hunk 3]
```diff
@@ -1297,6 +1310,32 @@ def updateStrategyPerformanceFee(
...
+    """
+    @notice
+        Change the custom strategy, default value is 0x0, when set to a non
+        null address, the vault will check the strategy health using this address.
+        If set to 0x0 it will use default checks.
+    @param strategy The Strategy to update.
+    @param _customCheck The contract that should perform the check, can be set to 0x0. 
+    """
+    assert msg.sender in [self.management, self.governance]
+    self.strategies[strategy].customCheck = _customCheck
```
Comment (fubuloubu) [inline contracts/Vault.vy:1337]: add `assert CustomHealthCheck(_customCheck).check(0, 0, strategy))` to make sure this function exists

File: contracts/Vault.vy [hunk 4]
```diff
@@ -1673,11 +1716,25 @@ def report(gain: uint256, loss: uint256, _debtPayment: uint256) -> uint256:
...
+    if self.strategies[msg.sender].enforceChangeLimit:
+        if self.strategies[msg.sender].customCheck != ZERO_ADDRESS:
+            assert(CustomHealthCheck(self.strategies[msg.sender].customCheck).check(gain, loss, msg.sender)) #dev: custom check
+        else:
+            totalDebt: uint256 = self.strategies[msg.sender].totalDebt
+
+            assert(gain <= ((totalDebt * self.strategies[msg.sender].profitLimitRatio) / MAX_BPS)) # dev: gain too high
+            assert(loss <= ((totalDebt * self.strategies[msg.sender].lossLimitRatio) / MAX_BPS)) # dev: loss too high
+    else:
+        self.strategies[msg.sender].enforceChangeLimit = True
```
Comment (fubuloubu) [inline contracts/Vault.vy:1731]: Add comment that this check can only be turned off for one report cycle

File: contracts/test/TestHealthCheck.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import {ICustomHealthCheck} from "../interfaces/ICustomHealthCheck.sol";
+
+contract TestHealthCheck is ICustomHealthCheck {
+    function check(
+        uint256 profit,
+        uint256 loss,
+        uint256 callerStrategy
+    ) external view override returns (bool) {
+        return true;
```
Comment (fubuloubu) [inline contracts/test/TestHealthCheck.sol:14]: Probably want to add the ability to change what this returns for testing

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**